test/integration/upload_test.lua

```diff
@@ -0,0 +1,161 @@
+local t = require('luatest')
+local g = t.group()
+
+local fio = require('fio')
+local fiber = require('fiber')
+local helpers = require('test.helper')
+
+g.before_all(function()
+    g.datadir = fio.tempdir()
+    g.upload_path_1 = fio.pathjoin(g.datadir, 'upload-1')
+    g.upload_path_2 = fio.pathjoin(g.datadir, 'upload-2')
+
+    g.cluster = helpers.Cluster:new({
+        datadir = g.datadir,
+        use_vshard = false,
+        server_command = helpers.entrypoint('srv_basic'),
+        cookie = helpers.random_cookie(),
+        replicasets = {{
+            alias = 'main',
+            uuid = helpers.uuid('a'),
+            roles = {},
+            servers = {
+                -- first upload group: s1 and s2
+                {env = {TARANTOOL_UPLOAD_PREFIX = g.upload_path_1}},
+                {env = {TARANTOOL_UPLOAD_PREFIX = g.upload_path_1}},
+                -- second upload group: s3
+                {env = {TARANTOOL_UPLOAD_PREFIX = g.upload_path_2}},
+            },
+        }},
+    })
+    g.cluster:start()
+
+    g.s1 = g.cluster:server('main-1')
+    g.s2 = g.cluster:server('main-2')
+    g.s3 = g.cluster:server('main-3')
+
+    -- This makes g.s1 to establish all pool connections.
+    -- Without it the first begining_failure test (with a SIGSTOP) is
+    -- slower and usually fails because the error message varies.
+    helpers.list_cluster_issues(g.s1)
+end)
+
+g.after_all(function()
+    g.cluster:stop()
+    fio.rmtree(g.cluster.datadir)
+end)
+
+function g.test_begining_failure()
+    g.s2.process:kill('STOP')
+
+    local future = g.s1.net_box:call(
+        'package.loaded.cartridge.config_patch_clusterwide',
+        {{['todo_list.txt'] = 'gotta go fast'}},
+        {is_async = true}
+    )
+
+    t.helpers.retrying({}, function()
+        t.assert(fio.listdir(g.upload_path_1)[1])
+        t.assert(fio.listdir(g.upload_path_2)[1])
+    end)
+
+    local res, err = unpack(future:wait_result())
+    t.assert_equals(res, nil)
+    t.assert_covers(err, {
+        class_name = 'NetboxCallError',
+        err = 'Timeout exceeded',
+    })
+
+    -- prefix should be cleaned up even if upload_begin fails
+    t.assert_equals(fio.listdir(g.upload_path_1), {})
+    t.assert_equals(fio.listdir(g.upload_path_2), {})
+
+    g.s2.process:kill('CONT')
+    -- give the instance some time to wake up after stop
+    fiber.sleep(0.3)
```

Comment:
Could you replace it with some wait_for?

Reply:
I was going to, but the predicate turned out to be super complicated. I need to wait when another async non-atomic (yielding) netbox call finishes on the server-side. I've decided that sleep is the shortest way. If the test turns out to be flaky I'll fix it later, but today it seems to be ok.

Reply:
ok